hazelcast/src/test/java/com/hazelcast/spi/addresslocator/AddressLocatorTest.java

```diff
@@ -0,0 +1,136 @@
+                .setEnabled(true)
+                .setClassname(AddressLocatorWithStaticProperties.class.getName())
+                .getProperties().setProperty("propName", "propValue");
+
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+
+        Hazelcast.newHazelcastInstance(config);
+        String property = AddressLocatorWithStaticProperties.properties.getProperty("propName");
+        assertEquals("propValue", property);
+    }
+
+    @Test
+    public void testSimpleAddressLocatorIsInjected() {
+        Config config = new Config();
+        config.getNetworkConfig().getAddressLocatorConfig()
+                .setEnabled(true)
+                .setClassname(SimpleAddressLocator.class.getName());
+
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+
+        InetSocketAddress inetSocketAddress = (InetSocketAddress) instance.getLocalEndpoint().getSocketAddress();
+
+        assertEquals(inetSocketAddress.getPort(), 9999);
+    }
+
+    @Test(expected = ConfigurationException.class)
+    public void testFailFastWhenNoMatchingConstructorIsFound() {
+        Config config = new Config();
+        config.getNetworkConfig().getAddressLocatorConfig()
+                .setEnabled(true)
+                .setClassname(SimpleAddressLocator.class.getName())
+                .getProperties().setProperty("foo", "bar"); // <-- this assumes AddressLocator has a constructor accepting Properties
+
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+
+        //this should fail
+        Hazelcast.newHazelcastInstance(config);
+    }
+
+    @Test
+    public void testPublicIPisTakenFromAddressLocator() {
+        Config config = new Config();
+        config.getNetworkConfig().getAddressLocatorConfig()
+                .setEnabled(true)
+                .setClassname(SimpleAddressLocator.class.getName());
+
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+        Member localMember = instance.getCluster().getLocalMember();
+
+        assertEquals("1.2.3.4", localMember.getAddress().getHost());
+    }
+
+    public static final class AddressLocatorWithStaticProperties implements AddressLocator {
+        public static Properties properties;
+
+        public AddressLocatorWithStaticProperties(Properties properties) {
+            this.properties = properties;
+        }
+
+        @Override
+        public InetSocketAddress getBindAddress() {
+            return new InetSocketAddress("localhost", 0);
+        }
+
+        @Override
+        public InetSocketAddress getPublicAddress() {
+            return new InetSocketAddress("localhost", 0);
+        }
+    }
+
+    public static final class SimpleAddressLocator implements AddressLocator {
```

Comment:
What happens on core side if concrete implementation returns non-routable IP?

Reply:
If the implementation returns a non-routable address for the bind address, the instance will fail to start. If it returns one for the public address, it fails to start the multicast service:
```
SEVERE: [123.123.123.123]:5701 [dev] [3.9-SNAPSHOT] Can't assign requested address
java.net.SocketException: Can't assign requested address
    at java.net.PlainDatagramSocketImpl.join(Native Method)
    at java.net.AbstractPlainDatagramSocketImpl.join(AbstractPlainDatagramSocketImpl.java:178)
    at java.net.MulticastSocket.joinGroup(MulticastSocket.java:323)
    at com.hazelcast.internal.cluster.impl.MulticastService.createMulticastService(MulticastService.java:137)
    at com.hazelcast.instance.Node.<init>(Node.java:219)
```